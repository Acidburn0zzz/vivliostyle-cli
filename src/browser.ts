import fs from 'fs';
import * as playwright from 'playwright-core';
import { registry } from 'playwright-core/lib/server';
import {
  beforeExitHandlers,
  logInfo,
  logSuccess,
  startLogging,
  stopLogging,
} from './util';

export async function launchBrowser(
  options?: playwright.LaunchOptions,
): Promise<playwright.Browser> {
  const browser = await playwright.chromium.launch({
    ...options,
  });
  beforeExitHandlers.push(() => {
    browser.close();
  });
  return browser;
}

export function getExecutableBrowserPath(): string {
  return playwright.chromium.executablePath();
}

export function checkBrowserAvailability(path: string): boolean {
  return fs.existsSync(path);
}

export function isPlaywrightExecutable(path: string): boolean {
  return registry.executables().some((exe) => exe.executablePath() === path);
}

export async function downloadBrowser(): Promise<string> {
  const executable = registry.findExecutable('chromium');
  logInfo('Rendering browser is not installed yet. Downloading now...');
  stopLogging();
  await registry.install([executable], false);
  logSuccess(`Successfully downloaded browser`);
  startLogging();
  return executable.executablePath()!;
}
