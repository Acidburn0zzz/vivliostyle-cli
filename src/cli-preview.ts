// cli-preview will be called when we uses `preview` subcommand

import path from 'path';
import program from 'commander';
import chalk from 'chalk';
import preview from './lib/preview';

program
  .name('vivliostyle preview')
  .description('launch preview server')
  .arguments('<input>')
  .option(
    '-r, --root <root_directory>',
    `specify assets root path (default directory of input file)`,
    undefined,
  )
  .option(
    '--no-sandbox',
    `launch chrome without sandbox (use this option to avoid ECONNREFUSED error)`,
  )
  .option(
    '--executable-chromium <path>',
    'specify a path of executable Chrome(Chromium) you installed',
  )
  .parse(process.argv);

if (program.args.length < 1) {
  program.help();
}

preview({
  input: path.resolve(process.cwd(), program.args[0]),
  rootDir: program.root,
  sandbox: program.sandbox,
  executableChromium: program.executableChromium,
}).catch((err) => {
  console.error(`${chalk.red.bold('Error:')} ${err.message}`);
  console.log(`
If you think this is a bug, please report at https://github.com/vivliostyle/vivliostyle-cli/issues`);
});
