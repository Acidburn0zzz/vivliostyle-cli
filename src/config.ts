import path from 'path';
import fs from 'fs';
import process from 'process';
import pkgUp from 'pkg-up';
import resolvePkg from 'resolve-pkg';
import { debug } from './util';
import { LoadMode } from './server';

export interface ParsedTheme {
  type: 'path' | 'uri';
  name: string;
  location: string;
}

export interface Entry {
  path: string;
  title?: string;
  theme?: string;
}

export interface VivliostyleConfig {
  title?: string;
  author?: string;
  theme: string; // undefined
  entry: string | string[] | Entry | Entry[];
  entryContext?: string;
  size?: string;
  format?: 'pdf';
  pressReady?: boolean;
  outDir?: string;
  outFile?: string; // output.pdf
  language?: string;
  cover?: string;
  toc?: boolean | string;

  timeout?: number;
  distDir?: string; // .vivliostyle
}

export interface CliFlags {
  configPath?: string;
  input?: string;
  size?: number | string;
  outFile?: string;
  outDir?: string;
  theme?: string;
  pressReady?: boolean;
  title?: string;
  author?: string;
  language?: string;
  entryContext?: string;
  verbose?: boolean;

  timeout?: number;
  loadMode?: LoadMode;
  sandbox?: boolean;
  executableChromium?: string;
}

const runningVivliostyleTimeout = 60 * 1000;
export function validateTimeout(val: string) {
  return Number.isFinite(+val) && +val > 0
    ? +val * 1000
    : runningVivliostyleTimeout;
}

export function contextResolve(
  context: string,
  loc: string | undefined,
): string | undefined {
  return loc && path.resolve(context, loc);
}

export function parseTheme(themeString: unknown): ParsedTheme | undefined {
  if (typeof themeString !== 'string') {
    return undefined;
  }

  // handle url
  if (/^https?:\/\//.test(themeString)) {
    return {
      type: 'uri',
      name: path.basename(themeString),
      location: themeString,
    };
  }

  const pkgRoot = resolvePkg(themeString, { cwd: process.cwd() });
  if (!pkgRoot) {
    throw new Error('package not found: ' + themeString);
  }

  // return bare .css path
  if (pkgRoot.endsWith('.css')) {
    return { type: 'path', name: path.basename(pkgRoot), location: pkgRoot };
  }

  // node_modules & local path
  const packageJson = JSON.parse(
    fs.readFileSync(path.join(pkgRoot, 'package.json'), 'utf8'),
  );

  const maybeCSS =
    packageJson?.vivliostyle?.theme?.style ||
    packageJson?.vivliostyle?.theme?.stylesheet ||
    packageJson.style ||
    packageJson.main;

  if (!maybeCSS || !maybeCSS.endsWith('.css')) {
    throw new Error('invalid css file: ' + maybeCSS);
  }

  return {
    type: 'path',
    name: `${packageJson.name.replace(/\//g, '-')}.css`,
    location: path.resolve(pkgRoot, maybeCSS),
  };
}

export function collectVivliostyleConfig(
  configPath: string,
): VivliostyleConfig | undefined {
  if (!fs.existsSync(configPath)) {
    return undefined;
  }
  const config = require(configPath) as VivliostyleConfig;
  return config;
}

export function getVivliostyleConfigPath(configPath?: string) {
  const cwd = process.cwd();
  return configPath
    ? path.resolve(cwd, configPath)
    : path.join(cwd, 'vivliostyle.config.js');
}

export async function mergeConfig<T extends CliFlags>(
  cliFlags: T,
  config: VivliostyleConfig | undefined,
  context: string,
) {
  const pkgJsonPath = await pkgUp();
  const pkgJson = pkgJsonPath ? require(pkgJsonPath) : undefined;

  // TODO: use dirname of cliFlags.input if it was given

  const projectTitle = cliFlags.title || config?.title || pkgJson?.name;
  if (!projectTitle) {
    throw new Error('title not defined');
  }
  const projectAuthor = cliFlags.author || config?.author || pkgJson?.author;

  const entryContextDir = path.resolve(
    cliFlags.input
      ? '.'
      : cliFlags.entryContext ||
          contextResolve(context, config?.entryContext) ||
          '.',
  );
  const distDir = path.resolve(
    contextResolve(context, config?.distDir) || '.vivliostyle',
  );
  const artifactDir = path.join(distDir, 'artifacts');

  const outDir = cliFlags.outDir || contextResolve(context, config?.outDir);
  const outFile = cliFlags.outFile || contextResolve(context, config?.outFile);

  if (outDir && outFile) {
    throw new Error('outDir and outFile cannot be combined.');
  }
  const outputPath = outDir
    ? path.resolve(outDir, `${projectTitle}.pdf`)
    : outFile || path.resolve('./output.pdf');
  debug('outputPath', outputPath);

  const language = config?.language || 'en';
  const size = cliFlags.size || config?.size;
  const toc =
    typeof config?.toc === 'string'
      ? contextResolve(context, config?.toc)!
      : config?.toc !== undefined
      ? config.toc
      : true;
  const pressReady = cliFlags.pressReady || config?.pressReady || false;
  const verbose = cliFlags.verbose || false;
  const timeout = cliFlags.timeout || config?.timeout || 3000;
  const sandbox = cliFlags.sandbox || true;
  const loadMode = cliFlags.loadMode || 'book';
  const executableChromium = cliFlags.executableChromium;

  const themeIndex: ParsedTheme[] = [];
  const projectTheme = cliFlags.theme || config?.theme;
  const rootTheme = projectTheme ? parseTheme(projectTheme) : undefined;
  if (rootTheme) {
    themeIndex.push(rootTheme);
  }

  const rawEntries = cliFlags.input
    ? [cliFlags.input]
    : config
    ? Array.isArray(config.entry)
      ? config.entry
      : config.entry
      ? [config.entry]
      : []
    : [];

  const parsedConfig = {
    entryContextDir,
    artifactDir,
    distDir,
    outputPath,
    rawEntries,
    themeIndex,
    toc,
    size,
    pressReady,
    projectTitle,
    projectAuthor,
    language,
    verbose,
    timeout,
    loadMode,
    sandbox,
    executableChromium,
  };

  debug(parsedConfig);

  return parsedConfig;
}
