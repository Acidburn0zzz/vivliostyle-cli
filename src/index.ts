export { build, BuildCliFlags } from './build';
export { init, InitCliFlags } from './init';
export { preview, PreviewCliFlags } from './preview';
