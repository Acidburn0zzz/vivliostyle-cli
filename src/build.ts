import chalk from 'chalk';
import terminalLink from 'terminal-link';
import path from 'upath';
import { getExecutableBrowserPath } from './browser';
import { checkOverwriteViolation, compile, copyAssets } from './builder';
import {
  CliFlags,
  collectVivliostyleConfig,
  mergeConfig,
  MergedConfig,
} from './config';
import { buildPDF, buildPDFWithContainer } from './pdf';
import { teardownServer } from './server';
import {
  checkContainerEnvironment,
  cwd,
  debug,
  log,
  startLogging,
  stopLogging,
} from './util';
import { exportWebPublication } from './webbook';

export interface BuildCliFlags extends CliFlags {
  output?: {
    output?: string;
    format?: string;
  }[];
  bypassedPdfBuilderOption?: string;
}

export async function getFullConfig(
  cliFlags: BuildCliFlags,
): Promise<MergedConfig[]> {
  const loadedConf = collectVivliostyleConfig(cliFlags);
  const { vivliostyleConfig, vivliostyleConfigPath } = loadedConf;
  const loadedCliFlags = loadedConf.cliFlags;

  const context = vivliostyleConfig ? path.dirname(vivliostyleConfigPath) : cwd;

  const configEntries: MergedConfig[] = [];
  for (const entry of vivliostyleConfig ?? [vivliostyleConfig]) {
    const config = await mergeConfig(loadedCliFlags, entry, context);
    checkUnsupportedOutputs(config);

    // check output path not to overwrite source files
    for (const target of config.outputs) {
      checkOverwriteViolation(config, target.path, target.format);
    }
    configEntries.push(config);
  }
  return configEntries;
}

export async function build(cliFlags: BuildCliFlags) {
  if (cliFlags.bypassedPdfBuilderOption) {
    const option = JSON.parse(cliFlags.bypassedPdfBuilderOption);
    // Host doesn't know browser path inside of container
    option.executableBrowser = getExecutableBrowserPath(
      option.browserType ?? 'chromium',
    );
    debug('bypassedPdfBuilderOption', option);

    startLogging();
    await buildPDF(option);
    // Stop remaining stream output and kill process
    stopLogging();
    return;
  }

  const isInContainer = checkContainerEnvironment();
  startLogging('Collecting build config');
  const configEntries = await getFullConfig(cliFlags);

  for (const config of configEntries) {
    // build artifacts
    if (config.manifestPath) {
      await compile(config);
      await copyAssets(config);
    }

    // generate files
    for (const target of config.outputs) {
      let output: string | null = null;
      if (target.format === 'pdf') {
        if (!isInContainer && target.renderMode === 'docker') {
          output = await buildPDFWithContainer({
            ...config,
            input: (config.manifestPath ??
              config.webbookEntryPath ??
              config.epubOpfPath) as string,
            target,
          });
        } else {
          output = await buildPDF({
            ...config,
            input: (config.manifestPath ??
              config.webbookEntryPath ??
              config.epubOpfPath) as string,
            target,
          });
        }
      } else if (target.format === 'webpub') {
        if (!config.manifestPath) {
          continue;
        }
        output = await exportWebPublication({
          ...config,
          input: config.workspaceDir,
          output: target.path,
        });
      }
      if (output) {
        const formattedOutput = chalk.bold.green(path.relative(cwd, output));
        log(
          `\n${terminalLink(formattedOutput, 'file://' + output, {
            fallback: () => formattedOutput,
          })} has been created.`,
        );
      }
    }

    teardownServer();
  }

  stopLogging('Built successfully.', 'ðŸŽ‰');
}

function checkUnsupportedOutputs({
  webbookEntryPath,
  epubOpfPath,
  outputs,
}: MergedConfig) {
  if (webbookEntryPath && outputs.some((t) => t.format === 'webpub')) {
    throw new Error(
      'Exporting webpub format from single HTML input is not supported.',
    );
  } else if (epubOpfPath && outputs.some((t) => t.format === 'webpub')) {
    throw new Error(
      'Exporting webpub format from EPUB or OPF file is not supported.',
    );
  }
}
