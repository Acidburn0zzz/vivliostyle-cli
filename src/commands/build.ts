import chalk from 'chalk';
import process from 'process';
import terminalLink from 'terminal-link';
import path from 'upath';
import { checkOverwriteViolation, compile, copyAssets } from '../builder';
import { collectVivliostyleConfig, mergeConfig } from '../config';
import { buildPDF } from '../pdf';
import { gracefulError, log, startLogging, stopLogging } from '../util';
import { exportWebbook } from '../webbook';
import { BuildCliFlags, setupBuildParserProgram } from './build.parser';

try {
  const program = setupBuildParserProgram();
  program.parse(process.argv);
  build({
    input: program.args?.[0],
    configPath: program.config,
    targets: program.targets,
    theme: program.theme,
    size: program.size,
    title: program.title,
    author: program.author,
    language: program.language,
    pressReady: program.pressReady,
    verbose: program.verbose,
    timeout: program.timeout,
    sandbox: program.sandbox,
    executableChromium: program.executableChromium,
  }).catch(gracefulError);
} catch (err) {
  gracefulError(err);
}

export default async function build(cliFlags: BuildCliFlags) {
  startLogging('Collecting build config');

  const loadedConf = collectVivliostyleConfig(cliFlags);
  const { vivliostyleConfig, vivliostyleConfigPath } = loadedConf;
  cliFlags = loadedConf.cliFlags;

  const context = vivliostyleConfig
    ? path.dirname(vivliostyleConfigPath)
    : process.cwd();

  const config = await mergeConfig(cliFlags, vivliostyleConfig, context);

  // check output path not to overwrite source files
  for (const target of config.outputs) {
    checkOverwriteViolation(config, target.path, target.format);
  }

  // build artifacts
  if (config.manifestPath) {
    await compile(config);
    await copyAssets(config);
  }

  // generate files
  for (const target of config.outputs) {
    let output: string | null = null;
    if (target.format === 'pdf') {
      output = await buildPDF({
        ...config,
        input: config.manifestPath ?? config.epubOpfPath,
        output: target.path,
      });
    } else if (target.format === 'webbook') {
      if (!config.manifestPath) {
        log(
          `\n${chalk.yellow(
            'Exporting webbook format from EPUB or OPF file is not supported. Skipping.',
          )}`,
        );
        continue;
      }
      output = await exportWebbook({
        ...config,
        input: config.workspaceDir,
        output: target.path,
      });
    } else if (target.format === 'pub-manifest') {
      // TODO
    }
    if (output) {
      const formattedOutput = chalk.bold.green(
        path.relative(process.cwd(), output),
      );
      log(
        `\n${terminalLink(formattedOutput, 'file://' + output, {
          fallback: () => formattedOutput,
        })} has been created.`,
      );
    }
  }

  stopLogging('Built successfully.', 'ðŸŽ‰');

  process.exit(0);
}
