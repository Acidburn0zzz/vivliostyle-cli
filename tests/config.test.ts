import shelljs from 'shelljs';
import path from 'upath';
import { setupBuildParserProgram } from '../src/commands/build.parser';
import { collectVivliostyleConfig, mergeConfig } from '../src/config';

const configFiles = ['valid.1', 'valid.2', 'valid.3', 'invalid.1'] as const;
const configFilePath = configFiles.reduce(
  (p, v) => ({
    ...p,
    [v]: path.resolve(__dirname, `fixtures/config/${v}.config.js`),
  }),
  {} as { [k in typeof configFiles[number]]: string },
);

const rootPath = path.resolve(__dirname, '..');
const getMergedConfig = async (args: string[]) => {
  const program = setupBuildParserProgram().parse([
    'vivliostyle',
    'build',
    ...args,
  ]);
  const {
    vivliostyleConfig,
    vivliostyleConfigPath,
    cliFlags,
  } = collectVivliostyleConfig({
    ...program.opts(),
    input: program.args?.[0],
    configPath: program.config,
    targets: program.targets,
  });
  const context = vivliostyleConfig
    ? path.dirname(vivliostyleConfigPath)
    : __dirname;
  const config = await mergeConfig(cliFlags, vivliostyleConfig, context);
  maskConfig(config);
  return config;
};

const maskConfig = (obj: any) => {
  Object.entries(obj).forEach(([k, v]) => {
    if (v && typeof v === 'object') {
      maskConfig(v);
    } else if (k === 'executableChromium') {
      obj[k] = '__EXECUTABLE_CHROMIUM_PATH__';
    } else if (typeof v === 'string') {
      obj[k] = v.replace(rootPath, '__WORKSPACE__');
    }
  });
};

afterAll(() => {
  shelljs.rm('-f', path.resolve(__dirname, 'fixtures/config/.vs-*'));
});

it('parse vivliostyle config', async () => {
  const validConfig1 = await getMergedConfig(['-c', configFilePath['valid.1']]);
  expect(validConfig1).toMatchSnapshot('valid.1.config.js');

  const validConfig2 = await getMergedConfig(['-c', configFilePath['valid.2']]);
  expect(validConfig2).toMatchSnapshot('valid.2.config.js');

  const validConfig3 = await getMergedConfig(['-c', configFilePath['valid.3']]);
  expect(validConfig3).toMatchSnapshot('valid.3.config.js');
});

it('override option by CLI command', async () => {
  const config = await getMergedConfig([
    '-c',
    configFilePath['valid.1'],
    '-o',
    'yuno.pdf',
    '-o',
    'yuno',
    '-f',
    'webbook',
    '-t',
    'https://myTheme.example.com',
    '-s',
    'JIS-B5',
    '--title',
    'myTitle',
    '--author',
    'myAuthor',
    '--language',
    'myLanguage',
    '--timeout',
    '42',
    '--executable-chromium',
    'myChromium',
  ]);
  expect(config).toMatchSnapshot('valid.1.config.js');
});

it('deny invalid config', () => {
  expect(
    getMergedConfig(['-c', configFilePath['invalid.1']]),
  ).rejects.toThrow();
});

it('Loads same config file on each way', async () => {
  const config1 = await getMergedConfig(['-c', configFilePath['valid.1']]);
  const config2 = await getMergedConfig([configFilePath['valid.1']]);
  expect(config1).toEqual(config2);
});

it('yields a config with single markdown', async () => {
  const config = await getMergedConfig([
    path.resolve(__dirname, 'fixtures/config/sample.md'),
  ]);
  expect(config.entries[0].target).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.sample\.md$/,
  );
  expect(config.manifestPath).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.manifest\.json$/,
  );
  delete config.manifestPath;
  delete config.entries[0].target;
  expect(config).toMatchSnapshot();
});

it('yields a config with single html', async () => {
  const config = await getMergedConfig([
    path.resolve(__dirname, 'fixtures/config/sample.html'),
  ]);
  expect(config.entries[0].target).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.sample\.html$/,
  );
  expect(config.manifestPath).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.manifest\.json$/,
  );
  delete config.manifestPath;
  delete config.entries[0].target;
  expect(config).toMatchSnapshot();
});

it('yields a config with single input and vivliostyle config', async () => {
  const config = await getMergedConfig([
    path.resolve(__dirname, 'fixtures/config/sample.md'),
    '-c',
    configFilePath['valid.1'],
  ]);
  expect(config.entries[0].target).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.sample\.md$/,
  );
  expect(config.manifestPath).toMatch(
    /^__WORKSPACE__\/tests\/fixtures\/config\/\.vs-.+\.manifest\.json$/,
  );
  delete config.manifestPath;
  delete config.entries[0].target;
  expect(config).toMatchSnapshot();
});
