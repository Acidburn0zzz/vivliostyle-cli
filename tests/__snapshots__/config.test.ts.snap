// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`override option by CLI command: valid.1.config.js 1`] = `
Object {
  "cover": "__WORKSPACE__/tests/fixtures/config/cover.png",
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
      "target": "__WORKSPACE__/tests/fixtures/config/manuscript.html",
      "theme": Object {
        "location": "https://myTheme.example.com",
        "name": "myTheme.example.com",
        "type": "uri",
      },
      "title": "myTitle",
      "type": "text/markdown",
    },
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
      "target": "__WORKSPACE__/tests/fixtures/config/manuscript.html",
      "theme": Object {
        "location": "__WORKSPACE__/tests/fixtures/config/theme.css",
        "name": "theme.css",
        "type": "file",
      },
      "title": "title",
      "type": "text/markdown",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": Array [],
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/publication.json",
    "format": "pub-manifest",
  },
  "language": "myLanguage",
  "manifestPath": "__WORKSPACE__/tests/fixtures/config/publication.json",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/yuno.pdf",
    },
    Object {
      "format": "webbook",
      "path": "__WORKSPACE__/yuno",
    },
  ],
  "pressReady": true,
  "projectAuthor": "myAuthor",
  "projectTitle": "myTitle",
  "sandbox": true,
  "size": Object {
    "format": "JIS-B5",
  },
  "themeIndexes": Array [
    Object {
      "location": "https://myTheme.example.com",
      "name": "myTheme.example.com",
      "type": "uri",
    },
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/theme.css",
      "name": "theme.css",
      "type": "file",
    },
  ],
  "timeout": 42000,
  "toc": "__WORKSPACE__/tests/fixtures/config/toc.html",
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`parse vivliostyle config: valid.1.config.js 1`] = `
Object {
  "cover": "__WORKSPACE__/tests/fixtures/config/cover.png",
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
      "target": "__WORKSPACE__/tests/fixtures/config/manuscript.html",
      "theme": Object {
        "location": "__WORKSPACE__/tests/fixtures/config/debug-theme",
        "name": "debug-theme",
        "style": "./theme.css",
        "type": "package",
      },
      "title": "title",
      "type": "text/markdown",
    },
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
      "target": "__WORKSPACE__/tests/fixtures/config/manuscript.html",
      "theme": Object {
        "location": "__WORKSPACE__/tests/fixtures/config/theme.css",
        "name": "theme.css",
        "type": "file",
      },
      "title": "title",
      "type": "text/markdown",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": Array [],
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/publication.json",
    "format": "pub-manifest",
  },
  "language": "language",
  "manifestPath": "__WORKSPACE__/tests/fixtures/config/publication.json",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output1.pdf",
    },
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output2.pdf",
    },
  ],
  "pressReady": true,
  "projectAuthor": "author",
  "projectTitle": "title",
  "sandbox": true,
  "size": Object {
    "format": "size",
  },
  "themeIndexes": Array [
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/debug-theme",
      "name": "debug-theme",
      "style": "./theme.css",
      "type": "package",
    },
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/theme.css",
      "name": "theme.css",
      "type": "file",
    },
  ],
  "timeout": 1,
  "toc": "__WORKSPACE__/tests/fixtures/config/toc.html",
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`parse vivliostyle config: valid.2.config.js 1`] = `
Object {
  "cover": undefined,
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
      "target": "__WORKSPACE__/tests/fixtures/config/manuscript.html",
      "theme": undefined,
      "title": "pkgName",
      "type": "text/markdown",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": Array [],
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/publication.json",
    "format": "pub-manifest",
  },
  "language": "en",
  "manifestPath": "__WORKSPACE__/tests/fixtures/config/publication.json",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output1.pdf",
    },
  ],
  "pressReady": false,
  "projectAuthor": "pkgAuthor",
  "projectTitle": "pkgName",
  "sandbox": true,
  "size": undefined,
  "themeIndexes": Array [],
  "timeout": 120000,
  "toc": false,
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`parse vivliostyle config: valid.3.config.js 1`] = `
Object {
  "cover": undefined,
  "entries": Array [],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": Array [],
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/publication.json",
    "format": "pub-manifest",
  },
  "language": "en",
  "manifestPath": "__WORKSPACE__/tests/fixtures/config/publication.json",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/example.pdf",
    },
  ],
  "pressReady": false,
  "projectAuthor": "pkgAuthor",
  "projectTitle": "example",
  "sandbox": true,
  "size": undefined,
  "themeIndexes": Array [],
  "timeout": 120000,
  "toc": false,
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`yields a config with single html 1`] = `
Object {
  "cover": undefined,
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/sample.html",
      "target": "__SNIP__",
      "theme": undefined,
      "title": "Miyako",
      "type": "text/html",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": "__SNIP__",
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/sample.html",
    "format": "webbook",
  },
  "language": "en",
  "manifestPath": "__SNIP__",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/output.pdf",
    },
  ],
  "pressReady": false,
  "projectAuthor": "",
  "projectTitle": "Miyako",
  "sandbox": true,
  "size": undefined,
  "themeIndexes": Array [],
  "timeout": 120000,
  "toc": false,
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`yields a config with single input and vivliostyle config 1`] = `
Object {
  "cover": "__WORKSPACE__/tests/fixtures/config/cover.png",
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/sample.md",
      "target": "__SNIP__",
      "theme": Object {
        "location": "__WORKSPACE__/tests/fixtures/config/debug-theme",
        "name": "debug-theme",
        "style": "./theme.css",
        "type": "package",
      },
      "title": "Yuno",
      "type": "text/markdown",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": "__SNIP__",
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/sample.md",
    "format": "markdown",
  },
  "language": "language",
  "manifestPath": "__SNIP__",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output1.pdf",
    },
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output2.pdf",
    },
  ],
  "pressReady": true,
  "projectAuthor": "author",
  "projectTitle": "title",
  "sandbox": true,
  "size": Object {
    "format": "size",
  },
  "themeIndexes": Array [
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/debug-theme",
      "name": "debug-theme",
      "style": "./theme.css",
      "type": "package",
    },
  ],
  "timeout": 1,
  "toc": "__WORKSPACE__/tests/fixtures/config/toc.html",
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;

exports[`yields a config with single markdown 1`] = `
Object {
  "cover": undefined,
  "entries": Array [
    Object {
      "source": "__WORKSPACE__/tests/fixtures/config/sample.md",
      "target": "__SNIP__",
      "theme": undefined,
      "title": "Yuno",
      "type": "text/markdown",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "executableChromium": "__EXECUTABLE_CHROMIUM_PATH__",
  "exportAliases": "__SNIP__",
  "includeAssets": Array [
    "**/*.png",
    "**/*.jpg",
    "**/*.jpeg",
    "**/*.svg",
    "**/*.gif",
    "**/*.webp",
    "**/*.apng",
    "**/*.ttf",
    "**/*.otf",
    "**/*.woff",
    "**/*.woff2",
  ],
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/sample.md",
    "format": "markdown",
  },
  "language": "en",
  "manifestPath": "__SNIP__",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/output.pdf",
    },
  ],
  "pressReady": false,
  "projectAuthor": "",
  "projectTitle": "Yuno",
  "sandbox": true,
  "size": undefined,
  "themeIndexes": Array [],
  "timeout": 120000,
  "toc": false,
  "verbose": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config",
}
`;
